lib/hamster/vector.rb

```diff
@@ -526,21 +526,9 @@ def shuffle
     # @return [Vector]
     def uniq(&block)
       array = self.to_a
-      if block_given?
-        if array.frozen?
-          self.class.new(array.uniq(&block).freeze)
-        elsif array.uniq!(&block) # returns nil if no changes were made
-          self.class.new(array.freeze)
-        else
-          self
-        end
-      elsif array.frozen?
-        self.class.new(array.uniq.freeze)
-      elsif array.uniq! # returns nil if no changes were made
-        self.class.new(array.freeze)
-      else
-        self
-      end
+      return self.class.new(array.uniq(&block).freeze) if array.frozen?
+      return self.class.new(array.freeze) if array.uniq!(&block)
```

Comment:
Not sure why freezing the input is necessary seeing as `#initialize` already does so.

To be honest this could entire method could probably be:
``` ruby
def uniq(&block)
  self.class.new(to_a.uniq(&block))
end
```

Though it's possible that this would be less efficient, I wonder what the threshold is and if (for smaller vectors), it would actually be faster.